loopin.go

```diff
@@ -794,85 +804,49 @@ func getTxFee(tx *wire.MsgTx, fee chainfee.SatPerKVByte) btcutil.Amount {
 func (s *loopInSwap) waitForSwapComplete(ctx context.Context,
-	htlcOutpoint *wire.OutPoint, htlcValue btcutil.Amount) error {
+	executorWg *sync.WaitGroup, htlcOutpoint *wire.OutPoint,
+	htlcValue btcutil.Amount) error {
 
 	// Register the htlc spend notification.
 	rpcCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	spendChan, spendErr, err := s.lnd.ChainNotifier.RegisterSpendNtfn(
-		rpcCtx, htlcOutpoint, s.htlc.PkScript, s.InitiationHeight,
-	)
-	if err != nil {
-		return fmt.Errorf("register spend ntfn: %v", err)
-	}
 
 	// Register for swap invoice updates.
 	rpcCtx, cancel = context.WithCancel(ctx)
 	defer cancel()
 	s.log.Infof("Subscribing to swap invoice %v", s.hash)
-	swapInvoiceChan, swapInvoiceErr, err := s.lnd.Invoices.SubscribeSingleInvoice(
+	invoices := s.lnd.Invoices
+	swapInvoiceChan, swapInvoiceErr, err := invoices.SubscribeSingleInvoice(
 		rpcCtx, s.hash,
 	)
 	if err != nil {
 		return fmt.Errorf("subscribe to swap invoice: %v", err)
 	}
 
-	// publishTxOnTimeout publishes the timeout tx if the contract has
-	// expired.
-	publishTxOnTimeout := func() (btcutil.Amount, error) {
-		if s.height >= s.LoopInContract.CltvExpiry {
-			return s.publishTimeoutTx(ctx, htlcOutpoint, htlcValue)
-		}
-
-		return 0, nil
-	}
-
-	// Check timeout at current height. After a restart we may want to
-	// publish the tx immediately.
-	var sweepFee btcutil.Amount
-	sweepFee, err = publishTxOnTimeout()
-	if err != nil {
-		return err
+	// We monitor the htlc spend by the server or the timeout sweep in a
+	// separate go routine since we don't want to block on them in order to
+	// continue with the next loop in once the invoice is settled.
+	progress := &swapProgress{
+		invoiceFinalized: false,
+		htlcKeyRevealed:  false,
 	}
+	executorWg.Add(1)
+	go func() {
+		defer executorWg.Done()
 
-	htlcSpend := false
-	invoiceFinalized := false
-	htlcKeyRevealed := false
-	for !htlcSpend || !invoiceFinalized {
-		select {
-		// Spend notification error.
-		case err := <-spendErr:
-			return err
-
-		// Receive block epochs and start publishing the timeout tx
-		// whenever possible.
-		case notification := <-s.blockEpochChan:
-			s.height = notification.(int32)
-
-			sweepFee, err = publishTxOnTimeout()
-			if err != nil {
-				return err
-			}
-
-			if invoiceFinalized && !htlcKeyRevealed {
-				htlcKeyRevealed = s.tryPushHtlcKey(ctx)
-			}
-
-		// The htlc spend is confirmed. Inspect the spending tx to
-		// determine the final swap state.
-		case spendDetails := <-spendChan:
-			s.log.Infof("Htlc spend by tx: %v",
-				spendDetails.SpenderTxHash)
-
-			err := s.processHtlcSpend(
-				ctx, spendDetails, htlcValue, sweepFee,
-			)
-			if err != nil {
-				return err
-			}
+		s.waitForSuccessOrTimeout(
+			ctx, htlcOutpoint, htlcValue, progress,
+		)
+	}()
 
-			htlcSpend = true
+	isFinalized := func() bool {
+		progress.Lock()
+		defer progress.Unlock()
+		return progress.invoiceFinalized
 
+	}
+	for !isFinalized() {
```

Comment:
Another way to do this loop could be to move it completely to the async goroutine and just have a single `invoiceFinalized` channel that we wait on here. That way the `htlcKeyRevealed` wouldn't be passed around and we can simplify the code by removing the `swapProgress` struct altogether.

Reply:
That sounds much simpler! I should have done this from the beginning :)